File: packages/turtlebot/scripts/controller.py
#!/usr/bin/env python3
"""
Turtlebot Controller - Python miniROS Package

Enhanced turtlebot control system with:
- Advanced movement patterns (forward, circle, square, figure-8)
- Odometry feedback processing
- Safety features and error handling
- Real-time status monitoring

Demonstrates miniROS Python bindings with robotics patterns.
"""

import mini_ros
import time
import math
import signal
import sys
from enum import Enum
from typing import Optional, Tuple

class MovementPattern(Enum):
    STOP = "stop"
    FORWARD = "forward"
    BACKWARD = "backward" 
    CIRCLE = "circle"
    SQUARE = "square"
    FIGURE8 = "figure8"
    SPIN = "spin"

class RobotPose:
    """Robot pose information"""
    def __init__(self):
        self.x: float = 0.0
        self.y: float = 0.0
        self.yaw: float = 0.0
        self.linear_vel: float = 0.0
        self.angular_vel: float = 0.0
        self.timestamp: float = time.time()

class TurtlebotController:
    """Enhanced Turtlebot Controller with safety features"""
    
    def __init__(self, node_name: str = "py_turtlebot"):
        print(f"🤖 Initializing {node_name}")
        print("📦 Package: turtlebot")
        print("📍 Executable: py_controller")
        
        # Initialize miniROS
        mini_ros.init()
        self.node = mini_ros.Node(node_name)
        
        # Publishers and subscribers
        self.cmd_vel_pub = self.node.create_publisher(
            mini_ros.TwistMessage, '/cmd_vel', 10
        )
        self.odom_sub = self.node.create_subscription(
            mini_ros.OdometryMessage, '/odom', self._odom_callback, 10
        )
        
        # State management
        self.current_pose = RobotPose()
        self.start_time = time.time()
        self.running = True
        
        # Safety limits
        self.max_linear_speed = 1.0   # m/s
        self.max_angular_speed = 2.0  # rad/s
        
        # Setup signal handlers for graceful shutdown
        signal.signal(signal.SIGINT, self._signal_handler)
        signal.signal(signal.SIGTERM, self._signal_handler)
        
        print("✅ Controller ready")
        
    def _signal_handler(self, signum, frame):
        """Handle shutdown signals gracefully"""
        print(f"\n📡 Received signal {signum}, shutting down...")
        self.running = False
        self.emergency_stop()
        
    def _odom_callback(self, odom_msg):
        """Process odometry feedback with improved error handling"""
        try:
            # Extract pose information
            self.current_pose.x = odom_msg.position[0]
            self.current_pose.y = odom_msg.position[1]
            self.current_pose.yaw = odom_msg.get_yaw()
            
            # Extract velocity information
            self.current_pose.linear_vel = odom_msg.velocity[0]
            self.current_pose.angular_vel = odom_msg.velocity[5]
            self.current_pose.timestamp = time.time()
            
            # Periodic status reporting
            if hasattr(self, '_last_odom_report'):
                if time.time() - self._last_odom_report > 2.0:
                    self._print_pose_status()
                    self._last_odom_report = time.time()
            else:
                self._last_odom_report = time.time()
                
        except Exception as e:
            print(f"⚠️ Error processing odometry: {e}")
    
    def _print_pose_status(self):
        """Print current robot pose status"""
        print(f"📍 Pose: x={self.current_pose.x:.2f}, y={self.current_pose.y:.2f}, "
              f"yaw={math.degrees(self.current_pose.yaw):.1f}°, "
              f"vel={self.current_pose.linear_vel:.2f}m/s")
    
    def publish_velocity(self, linear_x: float = 0.0, angular_z: float = 0.0) -> bool:
        """Send velocity command with safety checks"""
        # Apply safety limits
        safe_linear = max(-self.max_linear_speed, 
                         min(self.max_linear_speed, linear_x))
        safe_angular = max(-self.max_angular_speed, 
                          min(self.max_angular_speed, angular_z))
        
        # Create twist message
        twist = mini_ros.TwistMessage()
        twist.set_linear(safe_linear, 0.0, 0.0)
        twist.set_angular(0.0, 0.0, safe_angular)
        
        # Validate and publish
        if twist.validate():
            self.cmd_vel_pub.publish(twist)
            
            # Only log non-zero commands to reduce spam
            if abs(safe_linear) > 0.01 or abs(safe_angular) > 0.01:
                print(f"🎮 Velocity: linear={safe_linear:.2f}m/s, angular={safe_angular:.2f}rad/s")
            
            return True
        else:
            print(f"⚠️ Invalid velocity command: linear={linear_x:.2f}, angular={angular_z:.2f}")
            return False
    
    def execute_pattern(self, pattern: MovementPattern, duration: float) -> bool:
        """Execute movement pattern with improved timing and safety"""
        print(f"🤖 Executing {pattern.value} pattern for {duration:.1f}s")
        
        start_time = time.time()
        end_time = start_time + duration
        last_status = time.time()
        
        try:
            while time.time() < end_time and self.running:
                elapsed = time.time() - start_time
                
                # Calculate velocities based on pattern
                if pattern == MovementPattern.STOP:
                    linear, angular = 0.0, 0.0
                    
                elif pattern == MovementPattern.FORWARD:
                    linear, angular = 0.4, 0.0
                    
                elif pattern == MovementPattern.BACKWARD:
                    linear, angular = -0.3, 0.0
                    
                elif pattern == MovementPattern.CIRCLE:
                    linear, angular = 0.3, 0.8
                    
                elif pattern == MovementPattern.SQUARE:
                    # Square pattern: forward -> turn -> repeat
                    side_time = 3.0  # 3 seconds per side
                    turn_time = 1.6  # 1.6 seconds per turn (90°)
                    cycle_time = side_time + turn_time
                    
                    phase = (elapsed % cycle_time) / cycle_time
                    if phase < (side_time / cycle_time):
                        linear, angular = 0.3, 0.0  # Forward
                    else:
                        linear, angular = 0.0, math.pi / 2  # Turn 90°
                        
                elif pattern == MovementPattern.FIGURE8:
                    # Figure-8 using sinusoidal motion
                    frequency = 0.3  # Complete figure-8 every ~3.3 seconds
                    linear = 0.3
                    angular = 1.5 * math.sin(2 * math.pi * frequency * elapsed)
                    
                elif pattern == MovementPattern.SPIN:
                    linear, angular = 0.0, 1.5
                    
                else:
                    linear, angular = 0.0, 0.0
                
                # Publish velocity command
                if not self.publish_velocity(linear, angular):
                    print("⚠️ Failed to publish velocity, stopping pattern")
                    break
                
                # Print status every 3 seconds
                if time.time() - last_status > 3.0:
                    remaining = end_time - time.time()
                    print(f"📊 Pattern progress: {elapsed:.1f}s elapsed, {remaining:.1f}s remaining")
                    last_status = time.time()
                
                time.sleep(0.05)  # 20Hz control loop
                
        except KeyboardInterrupt:
            print("\n⏹️ Pattern interrupted by user")
            self.running = False
            
        finally:
            # Always stop the robot
            self.publish_velocity(0.0, 0.0)
        
        print("✅ Pattern execution completed")
        return True
    
    def publish_mock_odometry(self):
        """Publish mock odometry data for demonstration"""
        try:
            odom_pub = self.node.create_publisher(mini_ros.OdometryMessage, '/odom', 10)
            odom = mini_ros.OdometryMessage()
            odom.set_pose_2d(0.0, 0.0, 0.0)
            odom_pub.publish(odom)
            print("📡 Mock odometry published")
        except Exception as e:
            print(f"⚠️ Failed to publish mock odometry: {e}")
    
    def emergency_stop(self):
        """Emergency stop - immediately halt all motion"""
        print("🛑 Emergency stop activated!")
        for _ in range(3):  # Send multiple stop commands for safety
            self.publish_velocity(0.0, 0.0)
            time.sleep(0.1)
    
    def run_demo(self) -> bool:
        """Run comprehensive demonstration sequence"""
        print("🚀 Starting turtlebot Python controller demo")
        print("📋 Demo sequence - Advanced movement patterns:")
        
        try:
            # Publish initial mock odometry
            self.publish_mock_odometry()
            time.sleep(0.5)
            
            # Demo patterns with varying durations
            patterns = [
                (MovementPattern.FORWARD, 3.0, "1. Forward movement"),
                (MovementPattern.BACKWARD, 2.0, "2. Backward movement"),
                (MovementPattern.CIRCLE, 6.0, "3. Circular motion"),
                (MovementPattern.SQUARE, 18.0, "4. Square pattern (advanced)"),
                (MovementPattern.FIGURE8, 10.0, "5. Figure-8 pattern"),
                (MovementPattern.SPIN, 4.0, "6. Spin in place"),
            ]
            
            for i, (pattern, duration, description) in enumerate(patterns):
                if not self.running:
                    break
                    
                print(f"\n🔸 {description} ({duration}s)")
                
                if not self.execute_pattern(pattern, duration):
                    print(f"❌ Pattern {i+1} failed")
                    break
                
                # Brief pause between patterns
                if self.running and i < len(patterns) - 1:
                    print("⏸️ Pausing between patterns...")
                    time.sleep(1.0)
            
            if self.running:
                print("\n🎉 Python turtlebot controller demo completed successfully!")
                return True
            else:
                print("\n⏹️ Demo stopped by user")
                return False
                
        except Exception as e:
            print(f"❌ Demo failed with error: {e}")
            return False
        
        finally:
            self.emergency_stop()
    
    def cleanup(self):
        """Clean up resources"""
        print("🧹 Cleaning up resources...")
        self.emergency_stop()
        
        try:
            self.node.destroy_node()
            mini_ros.shutdown()
        except Exception as e:
            print(f"⚠️ Cleanup warning: {e}")
        
        print("🏁 Cleanup completed")

def main():
    """Main entry point"""
    print("=" * 50)
    print("🐍 miniROS Python: Turtlebot Controller")
    print("📦 Package: turtlebot")
    print("📍 Script: controller.py")
    print("=" * 50)
    
    controller = None
    try:
        # Create and run controller
        controller = TurtlebotController("py_turtlebot_ctrl")
        
        # Run the demonstration
        success = controller.run_demo()
        
        # Brief pause to see final messages
        if success:
            print("\n⏳ Demo completed, shutting down in 2 seconds...")
            time.sleep(2.0)
        
    except KeyboardInterrupt:
        print("\n⏹️ Interrupted by user (Ctrl+C)")
        
    except Exception as e:
        print(f"❌ Unexpected error: {e}")
        
    finally:
        if controller:
            controller.cleanup()
        
        print("🏁 Python controller finished")

if __name__ == "__main__":
    main() 

File: packages/turtlebot/src/controller.rs
//! # Turtlebot Controller
//!
//! A comprehensive turtlebot control system with:
//! - Movement pattern execution (forward, backward, circular, square)
//! - Odometry feedback processing
//! - Safety features with velocity limits
//! - Real-time status monitoring
//!
//! This demonstrates miniROS core robotics functionality with minimal complexity.

use mini_ros::prelude::*;
use mini_ros::types::{TwistMessage, OdometryMessage, Vector3, Point3D, Quaternion, Header, MiniRosMessage};
use std::f32::consts::PI;
use std::sync::{Arc, Mutex};
use std::time::Duration;
use tokio::time::{sleep, Instant};
use tracing::{info, warn, error};

/// Movement patterns for the turtlebot
#[derive(Debug, Clone)]
pub enum MovementPattern {
    Stop,
    Forward(f32),          // Forward velocity in m/s
    Backward(f32),         // Backward velocity in m/s
    TurnLeft(f32),         // Angular velocity in rad/s
    TurnRight(f32),        // Angular velocity in rad/s
    Circle(f32, f32),      // (linear_vel, angular_vel)
    Square(f32),           // Side length in meters
    Figure8(f32),          // Radius in meters
}

/// Robot pose information
#[derive(Debug, Clone)]
pub struct RobotPose {
    pub x: f32,
    pub y: f32,
    pub yaw: f32,
    pub linear_vel: f32,
    pub angular_vel: f32,
    pub timestamp: Instant,
}

impl Default for RobotPose {
    fn default() -> Self {
        Self {
            x: 0.0,
            y: 0.0,
            yaw: 0.0,
            linear_vel: 0.0,
            angular_vel: 0.0,
            timestamp: Instant::now(),
        }
    }
}

/// Turtlebot controller with improved state management
pub struct TurtlebotController {
    node: Node,
    cmd_vel_pub: Publisher<TwistMessage>,
    odom_sub: Subscriber<OdometryMessage>,
    current_pose: Arc<Mutex<RobotPose>>,
    movement_pattern: MovementPattern,
    start_time: Instant,
    max_linear_speed: f32,
    max_angular_speed: f32,
    running: bool,
}

impl TurtlebotController {
    /// Create a new turtlebot controller with safety limits
    pub async fn new(node_name: &str) -> Result<Self> {
        let mut node = Node::new(node_name)?;
        node.init().await?;

        // Create publishers and subscribers
        let cmd_vel_pub = node.create_publisher::<TwistMessage>("/cmd_vel").await?;
        let odom_sub = node.create_subscriber::<OdometryMessage>("/odom").await?;

        Ok(Self {
            node,
            cmd_vel_pub,
            odom_sub,
            current_pose: Arc::new(Mutex::new(RobotPose::default())),
            movement_pattern: MovementPattern::Stop,
            start_time: Instant::now(),
            max_linear_speed: 1.0,  // 1 m/s max
            max_angular_speed: 2.0, // 2 rad/s max
            running: true,
        })
    }

    /// Set safety limits
    pub fn set_speed_limits(mut self, max_linear: f32, max_angular: f32) -> Self {
        self.max_linear_speed = max_linear;
        self.max_angular_speed = max_angular;
        self
    }

    /// Set up odometry callback with improved error handling
    pub async fn setup_odometry_callback(&mut self) -> Result<()> {
        let pose_ref = self.current_pose.clone();

        self.odom_sub.on_message(move |odom: OdometryMessage| {
            let yaw = 2.0 * odom.pose.orientation.z.atan2(odom.pose.orientation.w);
            
            let new_pose = RobotPose {
                x: odom.pose.position.x,
                y: odom.pose.position.y,
                yaw,
                linear_vel: odom.twist.linear.x,
                angular_vel: odom.twist.angular.z,
                timestamp: Instant::now(),
            };

            if let Ok(mut pose) = pose_ref.lock() {
                *pose = new_pose;
            }

            info!(
                "📍 Robot pose: x={:.2}, y={:.2}, yaw={:.2}°, vel={:.2}m/s",
                odom.pose.position.x,
                odom.pose.position.y,
                yaw.to_degrees(),
                odom.twist.linear.x
            );
        })?;

        Ok(())
    }

    /// Publish velocity command with safety checks
    pub async fn publish_velocity(&self, linear: Vector3, angular: Vector3) -> Result<()> {
        // Apply safety limits
        let safe_linear = Vector3 {
            x: linear.x.clamp(-self.max_linear_speed, self.max_linear_speed),
            y: linear.y.clamp(-self.max_linear_speed, self.max_linear_speed),
            z: linear.z.clamp(-self.max_linear_speed, self.max_linear_speed),
        };
        
        let safe_angular = Vector3 {
            x: angular.x.clamp(-self.max_angular_speed, self.max_angular_speed),
            y: angular.y.clamp(-self.max_angular_speed, self.max_angular_speed),
            z: angular.z.clamp(-self.max_angular_speed, self.max_angular_speed),
        };

        let twist = TwistMessage { 
            linear: safe_linear, 
            angular: safe_angular 
        };
        
        // Validate command
        if let Err(e) = twist.validate() {
            warn!("⚠️ Invalid twist command: {}", e);
            return Ok(());
        }
        
        self.cmd_vel_pub.publish(&twist).await?;
        
        if twist.linear.x != 0.0 || twist.angular.z != 0.0 {
            info!(
                "🎮 Velocity: linear={:.2}m/s, angular={:.2}rad/s",
                twist.linear.x, twist.angular.z
            );
        }
        
        Ok(())
    }

    /// Execute movement pattern with improved timing
    pub async fn execute_pattern(&mut self, pattern: MovementPattern, duration_sec: f64) -> Result<()> {
        self.movement_pattern = pattern.clone();
        self.start_time = Instant::now();
        
        info!("🤖 Starting movement pattern: {:?} for {:.1}s", pattern, duration_sec);
        
        let end_time = Instant::now() + Duration::from_secs_f64(duration_sec);
        let mut last_status = Instant::now();
        
        while Instant::now() < end_time && self.running {
            // Calculate current phase for time-based patterns
            let elapsed = self.start_time.elapsed().as_secs_f32();
            
            match &self.movement_pattern {
                MovementPattern::Stop => {
                    self.publish_velocity(
                        Vector3 { x: 0.0, y: 0.0, z: 0.0 },
                        Vector3 { x: 0.0, y: 0.0, z: 0.0 },
                    ).await?;
                }
                
                MovementPattern::Forward(speed) => {
                    self.publish_velocity(
                        Vector3 { x: *speed, y: 0.0, z: 0.0 },
                        Vector3 { x: 0.0, y: 0.0, z: 0.0 },
                    ).await?;
                }
                
                MovementPattern::Backward(speed) => {
                    self.publish_velocity(
                        Vector3 { x: -speed, y: 0.0, z: 0.0 },
                        Vector3 { x: 0.0, y: 0.0, z: 0.0 },
                    ).await?;
                }
                
                MovementPattern::TurnLeft(angular_speed) => {
                    self.publish_velocity(
                        Vector3 { x: 0.0, y: 0.0, z: 0.0 },
                        Vector3 { x: 0.0, y: 0.0, z: *angular_speed },
                    ).await?;
                }
                
                MovementPattern::TurnRight(angular_speed) => {
                    self.publish_velocity(
                        Vector3 { x: 0.0, y: 0.0, z: 0.0 },
                        Vector3 { x: 0.0, y: 0.0, z: -angular_speed },
                    ).await?;
                }
                
                MovementPattern::Circle(linear_vel, angular_vel) => {
                    self.publish_velocity(
                        Vector3 { x: *linear_vel, y: 0.0, z: 0.0 },
                        Vector3 { x: 0.0, y: 0.0, z: *angular_vel },
                    ).await?;
                }
                
                MovementPattern::Square(side_length) => {
                    // Improved square pattern with proper timing
                    let side_time = side_length / 0.3; // 0.3 m/s forward speed
                    let turn_time = (PI / 2.0) / 1.0;  // 1.0 rad/s turn speed
                    let cycle_time = side_time + turn_time;
                    
                    let phase = (elapsed % cycle_time) / cycle_time;
                    
                    if phase < (side_time / cycle_time) {
                        // Move forward
                        self.publish_velocity(
                            Vector3 { x: 0.3, y: 0.0, z: 0.0 },
                            Vector3 { x: 0.0, y: 0.0, z: 0.0 },
                        ).await?;
                    } else {
                        // Turn 90 degrees
                        self.publish_velocity(
                            Vector3 { x: 0.0, y: 0.0, z: 0.0 },
                            Vector3 { x: 0.0, y: 0.0, z: 1.0 },
                        ).await?;
                    }
                }
                
                MovementPattern::Figure8(radius) => {
                    // Figure-8 pattern using sinusoidal motion
                    let frequency = 0.5; // Complete figure-8 every 2 seconds
                    let linear_speed = 0.3;
                    let angular_speed = (2.0 * PI * frequency * (elapsed * frequency).sin()) / radius;
                    
                    self.publish_velocity(
                        Vector3 { x: linear_speed, y: 0.0, z: 0.0 },
                        Vector3 { x: 0.0, y: 0.0, z: angular_speed },
                    ).await?;
                }
            }
            
            // Print status every 2 seconds
            if last_status.elapsed() > Duration::from_secs(2) {
                self.print_status().await;
                last_status = Instant::now();
            }
            
            sleep(Duration::from_millis(50)).await; // 20Hz control loop
        }
        
        // Stop the robot
        self.publish_velocity(
            Vector3 { x: 0.0, y: 0.0, z: 0.0 },
            Vector3 { x: 0.0, y: 0.0, z: 0.0 },
        ).await?;
        
        info!("✅ Movement pattern completed");
        Ok(())
    }

    /// Print current status
    async fn print_status(&self) {
        if let Ok(pose) = self.current_pose.lock() {
            info!(
                "📊 Status: pos=({:.2}, {:.2}), yaw={:.1}°, speeds=({:.2}, {:.2})",
                pose.x, pose.y, pose.yaw.to_degrees(), pose.linear_vel, pose.angular_vel
            );
        }
    }

    /// Stop the robot immediately
    pub async fn emergency_stop(&mut self) -> Result<()> {
        self.running = false;
        self.publish_velocity(
            Vector3 { x: 0.0, y: 0.0, z: 0.0 },
            Vector3 { x: 0.0, y: 0.0, z: 0.0 },
        ).await?;
        warn!("🛑 Emergency stop activated!");
        Ok(())
    }

    /// Run comprehensive demonstration sequence
    pub async fn run_demo(&mut self) -> Result<()> {
        info!("🚀 Starting turtlebot controller demo");
        
        // Setup odometry callback
        self.setup_odometry_callback().await?;
        
        // Publish initial mock odometry for demonstration
        self.publish_mock_odometry().await?;
        
        sleep(Duration::from_millis(1000)).await;
        
        // Execute demonstration patterns
        info!("📋 Demo sequence - Turtlebot movement patterns:");
        
        // 1. Forward movement
        info!("🔸 1. Forward movement (3s)");
        self.execute_pattern(MovementPattern::Forward(0.4), 3.0).await?;
        sleep(Duration::from_secs(1)).await;
        
        // 2. Backward movement
        info!("🔸 2. Backward movement (2s)");
        self.execute_pattern(MovementPattern::Backward(0.3), 2.0).await?;
        sleep(Duration::from_secs(1)).await;
        
        // 3. Circular motion
        info!("🔸 3. Circular motion (5s)");
        self.execute_pattern(MovementPattern::Circle(0.3, 0.8), 5.0).await?;
        sleep(Duration::from_secs(1)).await;
        
        // 4. Square pattern
        info!("🔸 4. Square pattern (15s)");
        self.execute_pattern(MovementPattern::Square(1.0), 15.0).await?;
        sleep(Duration::from_secs(1)).await;
        
        // 5. Figure-8 pattern
        info!("🔸 5. Figure-8 pattern (10s)");
        self.execute_pattern(MovementPattern::Figure8(0.5), 10.0).await?;
        sleep(Duration::from_secs(1)).await;
        
        // 6. Spin in place
        info!("🔸 6. Spin in place (4s)");
        self.execute_pattern(MovementPattern::TurnLeft(1.5), 4.0).await?;
        
        info!("🎉 Turtlebot controller demo completed!");
        Ok(())
    }

    /// Publish mock odometry data for demonstration
    async fn publish_mock_odometry(&mut self) -> Result<()> {
        let odom_pub = self.node.create_publisher::<OdometryMessage>("/odom").await?;
        
        let current_time = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap_or_default()
            .as_nanos() as i64;
        
        let odom = OdometryMessage {
            pose: mini_ros::types::PoseMessage {
                position: Point3D { x: 0.0, y: 0.0, z: 0.0 },
                orientation: Quaternion { x: 0.0, y: 0.0, z: 0.0, w: 1.0 },
                header: Header {
                    timestamp: current_time,
                    frame_id: "odom".to_string(),
                },
            },
            twist: TwistMessage {
                linear: Vector3 { x: 0.0, y: 0.0, z: 0.0 },
                angular: Vector3 { x: 0.0, y: 0.0, z: 0.0 },
            },
            header: Header {
                timestamp: current_time,
                frame_id: "odom".to_string(),
            },
        };
        
        odom_pub.publish(&odom).await?;
        info!("📡 Published initial odometry data");
        
        Ok(())
    }
}

#[tokio::main]
async fn main() -> Result<()> {
    // Initialize tracing with better formatting
    tracing_subscriber::fmt()
        .with_env_filter("info")
        .init();

    info!("=== miniROS Turtlebot Controller ===");
    info!("🤖 Package: turtlebot");
    info!("📍 Executable: controller");

    // Create turtlebot controller with safety limits
    let mut controller = TurtlebotController::new("turtlebot_controller")
        .await?
        .set_speed_limits(0.8, 2.0); // Reasonable speed limits
    
    // Handle Ctrl+C gracefully
    let running = Arc::new(Mutex::new(true));
    let running_clone = running.clone();
    
    tokio::spawn(async move {
        tokio::signal::ctrl_c().await.unwrap();
        info!("📡 Received Ctrl+C, stopping...");
        *running_clone.lock().unwrap() = false;
    });
    
    // Run the demonstration
    match controller.run_demo().await {
        Ok(()) => info!("✅ Demo completed successfully"),
        Err(e) => error!("❌ Demo failed: {}", e),
    }
    
    // Ensure robot is stopped
    controller.emergency_stop().await?;
    
    // Keep node alive briefly to process final messages
    sleep(Duration::from_secs(1)).await;

    info!("🏁 Turtlebot controller finished");
    Ok(())
} 